### PR TITLE
Compatibility with Aeson for unary sum types

### PR DESCRIPTION
**Pull request by alexlegg**
I ran into an issue when generating encodings for unary sum types and using the generated encodings with Aeson.

Say we have this type:

`Data AType = AType Text`

Aeson encodes this to just a string, whereas elm-bridge constructs objects like `{"AType", "string"}`.

I also added some missing imports (Dict and Set) to the generated elm code.

If you want elm-bridge to maintain compatibility then I hope this may be helpful.

### CODE REVIEW

**Comment by agrafix** [comment]
Thanks, that sounds like a reasonable change to me. What do you think @bartavelle ?

**Comment by bartavelle** [comment]
I'll take a good look during the week.

**Comment by bartavelle** [comment]
This looks good, but I am afraid this might cause incompatibilities with aeson <= 0.9. I *think* (might be wrong, didn't check much) that this is triggered by the new `unwrapUnaryRecords` option in aeson 0.11. If that is the case, then it should be stored in the options structure during derivation and applied conditionally.

**Comment by agrafix** [comment]
Yes, we should either apply that conditionally, or alternatively drop `aeson <= 0.9` support.

**Comment by bartavelle** [comment]
Even with ditching 0.9 the problem will remain, as this is configurable behavior.

**Comment by alexlegg** [comment]
I'll have a look at the aeson code but when I flipped that flag the
behaviour that I described remains (that was the first thing I tried).
However, there may be a bug in aeson?

I think that flag refers to types like:
`Data Rec = Rec { field :: Text }`

**Comment by alexlegg** [comment]
Hmm, I've introduced a quirk that you might not like. When aeson encodes the string `"abc"` into a string, it merely becomes `"abc"`. When elm decodes `"abc"` to a string it will be an error. Instead, elm expects to see quotations inside the string `""abc""` (because it is expecting JSON, not a plain string).

Normally this isn't really a problem since you don't ever really need to decode a string. However, since `Data AType = AType Text` now encodes to a string you can get a confusing error when decoding it.

This isn't a problem if `AType` is nested within another type, only when you run `decodeString` with `jsonDecAType`.

Let me know what you think.

**Comment by bartavelle** [comment]
I just pushed some new test cases for the end-to-end tests that support this use case. They fail with the current tree and with your modification with the following error:

    Cannot use 'in' operator to search for 'qux' in -10

This is because of an error somewhere in the JSON stuff that decodes plain numbers into numbers, and then uses string comparison. Anyway, this can be fixed manually for now to check what is wrong with the code:

    mkdir /tmp/lala
    stack test
    .stack-work/dist/x86_64-linux/Cabal-1.22.5.0/build/end-to-end-tests/end-to-end-tests /tmp/lala/lala.elm
    cd /tmp/lala
    elm package install deadfoxygrandpa/elm-test
    elm package install bartavelle/json-helpers
    [edit lala.elm, see below]
    elm make lala.elm

In order to edit the `lala.elm`, you need to go transform `assertEqualHack` into `assertEqual` for all lines after the `simpleEncode01` function declaration (should be line 1052).

**Comment by bartavelle** [comment]
Hopefully you should then see the list of failing tests, and the corresponding `aeson` options!

**Comment by bartavelle** [comment]
Your patch fixes a lot of failing cases, but is not OK for some cases. This PR is a good step forward. Do you have time to take a look at the failing cases ?

**Comment by bartavelle** [comment]
I just pushed a fix to make unnecessary the modification of `lala.elm`.

**Comment by alexlegg** [comment]
Sure, I'll take a look tomorrow.

**Comment by alexlegg** [comment]
Ok I've fixed those cases. It looks like `unwrapUnaryRecords` was not handled at all previously.

**Comment by bartavelle** [comment]
It looks good to me!

File: test/EndToEnd.hs [hunk 1]
```diff
@@ -231,7 +231,7 @@ mkDecodeTest pred prefix num elems = unlines (
     ++ ["  ]"]
     )
   where
-      mktest (n,e) = pfix ++ "test \"" ++ show n ++ "\" (assertEqual (Json.Decode.decodeString (jsonDec" ++ pred ++ num ++ " (Json.Decode.list Json.Decode.int)) " ++ encoded ++ ") (Ok (" ++ pretty ++ ")))"
+      mktest (n,e) = pfix ++ "test \"" ++ show n ++ "\" (assertEqual (Ok (" ++ pretty ++ ")) (Json.Decode.decodeString (jsonDec" ++ pred ++ num ++ " (Json.Decode.list Json.Decode.int)) " ++ encoded ++ "))"
```
Comment (bartavelle) [inline test/EndToEnd.hs]: Nice catch :)

**Comment by agrafix** [comment]
Looks good to me too! Before merging: Could you please update the changelog? :-)

**Comment by alexlegg** [comment]
Sure, I just added a line describing the change. I'll defer any version number changes to you though.

**Pull request merged**